rfcs/20200619-dropout/README.md

```diff
@@ -0,0 +1,425 @@
+only one Bernoulli distribution) and usage one api call vs two api calls.
+The main pros of standalone RNG is flexibility of this solution, but it's
+common usability is questionable.
+
+### Random number generation
+
+The main challenge of the implementation is to create random number generation
+in one dnn. Such implementation should have the following properties:
+- efficiency - implementation in one dnn should be at least as efficient as
+viRngBernoulli from VSL and optimized for usage of multithreading and sse4.1,
+avx2, avx512
+- testability - by setting the same seed we should be able to create exactly the
+same mask using benchdnn and one dnn optimized implementation
+
+In case of calculation reference output in tests by benchdnn, mask generation
+with bernoulli distribution can be achieved by using C++ standard library. Since
+C++11 bernoulli_distribution is part of C++ standard library, by utilizing
+discard function and omp we can achieve mulithreaded effect as in example below:
+
+**C++ reference single thread**
+```cpp
+#include <vector>
+#include <random>
+#include <algorithm>
+#include <experimental/iterator>
+#include <iostream>
+#include <cmath>
+
+int main()
+{
+    const int seed = 10;
+    const float p = 0.5;
+    const std::size_t mask_size = 25;
+    std::vector<bool> mask(mask_size);
+
+    std::mt19937 engine(seed);
+    std::bernoulli_distribution gen(p);
+    auto generator = [&](...) -> bool { return gen(engine); };
+    std::transform(mask.begin(), mask.end(), mask.begin(), generator);
+
+    std::copy(mask.begin(), mask.end(), std::experimental::make_ostream_joiner(std::cout, ", "));
+}
+```
+
+**C++ reference multithreaded**
+
+```cpp
+#include <vector>
+#include <random>
+#include <algorithm>
+#include <experimental/iterator>
+#include <iostream>
+#include <cmath>
+#include <omp.h>
+
+template <typename T>
+void generate(size_t min, size_t max, T & collection, int seed, float p)
+{
+    std::mt19937 engine(seed);
+    std::bernoulli_distribution gen(p);
+
+    auto b = std::numeric_limits<long double>::digits;
+    auto R = engine.max() - engine.min() + 1;
+    auto discard_value = std::max(1.0, std::ceil(b/std::log2(R)));
+    engine.discard(min*discard_value);
+    const auto generator = [&](...) -> bool { return gen(engine); };
+    std::transform(collection.begin()+min, collection.begin()+max, collection.begin()+min, generator);
+}
+
+int main()
+{
+    const int seed = 10;
+    const float p = 0.5;
+    const std::size_t mask_size = 25;
+    std::vector<bool> mask(mask_size);
+
+    #pragma omp parallel num_threads(5)
+    {
+        int tid = omp_get_thread_num();
+
+        generate((tid*5), (tid*5) + 5, mask, seed, p);
+    }
+
+    std::copy(mask.begin(), mask.end(), std::experimental::make_ostream_joiner(std::cout, ", "));
+}
+```
+
+Both single threaded and multithreaded versions will output exactly the same
+value. For example Bernoulli distribution implementation for gcc is defined
```

Comment:
I totally agree with you.  
1. Production kernel will be implemented if possible as vectorized implementation in jit assembly described in chapter "Random number generation - vectorized implementation".
2. C++ implementation will be used mostly in benchdnn reference output calculations. Regarding different glibc implementations I'm aware of that and it is valid point. I just wanted to emphasize that writing distribution which is consistent with concept of standard library distribution is quite easy. To avoid issues with differen glibc writing internal implementation consistent with concepts from std will be needed.